src/hotspot/share/gc/shenandoah/heuristics/shenandoahAdaptiveHeuristics.cpp

```diff
@@ -123,12 +123,18 @@ void ShenandoahAdaptiveHeuristics::choose_collection_set_from_regiondata(Shenand
             add_region = true;
             old_cur_cset = new_cset;
           }
-        } else if (r->age() >= InitialTenuringThreshold) {
+        } else if (cset->is_preselected(r->index())) {
+          assert(r->age() >= InitialTenuringThreshold, "Preselected regions must have tenure age");
           // Entire region will be promoted, This region does not impact young-gen or old-gen evacuation reserve.
           // This region has been pre-selected and its impact on promotion reserve is already accounted for.
           add_region = true;
           cur_young_garbage += r->garbage();
-        } else {
+          // Since all live data in this region is being evacuated from young-gen, it is as if this memory
+          // is garbage insofar as young-gen is concerned.  Counting this as garbage reduces the need to
+          // reclaim highly utilized young-gen regions just for the sake of finding min_garbage to reclaim
+          // within youn-gen memory
+          cur_young_garbage += r->get_live_data_bytes();
```

Comment:
Isn't `r->garbage() + r->get_live_data_bytes()` equal to `r->used()`?

I am not sure I understand the reasoning here. I would much rather we named the `cur_young_garbage` variable more correctly if it is counting the amount of young regions that are in some sense being "removed from the young generation" (if indeed that's how to interpret it).

Reply:
Might be a good idea to describe at line 97, what `cur_young_garbage` will track as you process regions and build the collection set. Similarly, young_curset etc.

I see three separate outer if/else statements covering respectively the cases of generational/global, generational/not global, and non-generational cases. Would it read more easily and make for more easily maintained code to refactor into 3 separate cset builders called for each of the three cases, if necessary from the top level method (but even better if called from different contexts to specialize early hoisting the check on the type of collection up into the highest caller that has the information, rather than having all of them in one single, somewhat long and somewhat unwieldy method?

Reply:
Thanks for these suggestions.  Adding a comment and "folding" those two addition operations into one.

Reply:
Not refactoring the body of method for now, but that can be done at a future time.